### DISCUSSION TITLE
Connecting Vercel serverless functions to SQL database

### DISCUSSION THREAD

**Discussion by DavidChouinard** [vercel/community#993] [Help] [answered]
What's the best practice for connecting to an SQL database from Vercel serverless functions? I've seen a few options commonly mentioned:

 - Create a new direct database connection in the serverless function. This has important drawbacks:
   - **Connection pooling**: every new invocation of a serverless function would create its database connection, which could quickly overwhelm the database
   - **Security**: the database has to be publicly exposed since Vercel [doesn't support static IPs](https://vercel.com/guides/can-i-get-a-fixed-ip-address) or VPC peering. This unfortunately is deal-breaker for any security-sensitive application (fintech, healthcare, education, etc.) and SOC 2 compliance
 - Add an intermediary service that receives HTTP requests and proxies it to the database
   - My understanding is that this is a common thing people do? How does this work?
 - Use a vendor-specific solution, like the [Prisma Data Proxy](https://www.prisma.io/data-platform/proxy) product (requires using using the Prisma ORM) or AWS Aurora [Data API](https://docs.aws.amazon.com/AmazonRDS/latest/AuroraUserGuide/data-api.html) (essentially an out-of-the-box version of the second option, now deprecated)

Trying to understand what the "correct" or "best practice" solution to this problem is — have others deployed solutions they're satisfied with?

**Comment by leerob** [comment, marked as answer]
Hey @DavidChouinard! 👋 

This is mostly accurate, a few additional thoughts:

- Re: pooling, some database providers, like Planetscale, can handle [_a lot_](https://planetscale.com/blog/one-million-connections) of connections – where the pooling problem isn't an issue.
- Re: securely connecting to databases, happy to talk more about that with you and explore some options we have (lee@vercel.com).
- Re: HTTP proxies, yes you could roll your own through API gateway for example.
- Re: HTTP solutions, yes Prisma/Data API are both common solutions. Another similar "database API as a service" is [Supabase](https://supabase.com/).

I can't say which is the "best" because it depends on the application, but this is a good overview of some of the solutions!

**Comment by DavidChouinard** [comment]
Awesome very helpful! Seems like using a provider like Planet Scale that can handle lots of connections is the best way to go (Aurora Serverless v2 seems to fit the bill as well for the AWS ecosystem).

That leaves the security problem to handle — unfortunately solutions like setting an API gateway would require rewriting every instance where we query the db to use HTTP calls instead of an SQL connection :(

Been speaking to our AE and looks like the upcoming Vercel Connect feature will also be HTTP only? (no SQL traffic)

**Reply by swijckmans** [reply]
We are looking into supporting any form of TCP requests, this include Redis and Postgres.

**Reply by DavidChouinard** [reply]
Exciting!

**Comment by mikecfisher** [comment]
What's the latest on securely connecting to your database on a Vercel-hosted app? It seems that unless you are on an enterprise plan, you have to open your database to the internet to use Vercel. 

I know Vercel Enterprise has their new Secure Compute product and supports VPC peering; however, the price for that is in the "low four figures a month" which would be dramatically more expensive than our current AWS bill. 

I'm rebuilding a legacy app on NextJS and trying to host it on Vercel however, my team is against it because of the requirement to open our RDS instance to the open internet.
